(SIMP-8601) Add and document role-based formatting

op-ct opened the pull request:
This patch implements simp-specific sphinx/ReST/docutils roles + CSS
styles to standardize the formatting in our technical docs (at least,
compared to spamming double-backticks everywhere).  

To show them in action, I've:

* converted the relevant double-backticks in the
  "[Version-Specific Upgrade Instructions][0]" to use the new roles.
* added a Developer-facing [style guide for using the roles][1] in the
  Contribution Procedure.
* added a user-facing [typographical conventions documentation][2] in
  the "Getting Started" Guide.

The tables in all documentation pulls from the same data source to
keep things DRY.

Notes:

* You can see the rendered formatting in action at the links above.
* The colors, formatting, number of roles are just for demonstration
  purposes; we can change them to anything we want.

[SIMP-8602] #close 
[SIMP-8603] #close 
[SIMP-8604] #close 


[SIMP-8602]: https://simp-project.atlassian.net/browse/SIMP-8602
[SIMP-8603]: https://simp-project.atlassian.net/browse/SIMP-8603
[SIMP-8604]: https://simp-project.atlassian.net/browse/SIMP-8604

[0]: https://simp.readthedocs.io/en/simp-maint-example-role-formatting/user_guide/Upgrade_SIMP/Version_Specific_Upgrade_Instructions.html
[1]: https://simp.readthedocs.io/en/simp-maint-example-role-formatting/contributors_guide/Documentation.html
[2]: https://simp.readthedocs.io/en/simp-maint-example-role-formatting/getting_started_guide/Introduction.html#conventions-used-in-this-documentation

trevor-vaughan commented:
This is a LOT easier to deal with. How about we move the 'Sample' bits into the Developer Guide and get this through?

It would be great to have it all done at once but I don't know that this is realistic (either to do or to review) all at the same time so it may need to be incremental change starting with the User's Guide.

judyj commented on docs/user_guide/Upgrade_SIMP/Version_Maps/6.4.0_6.5.0.inc:
```diff
@@ -74,33 +125,35 @@ In Puppet code:
 
-Set ``svckill::mode`` to ``warning``
-""""""""""""""""""""""""""""""""""""
+Set :code:`svckill::mode` to ``warning``
```
should the `` symbol be defined above?

op-ct replied:
I'd considered that, too—but the existing charts described the roles, and it seemed overkill to make a table just for one row of general syntax.

However, if it helps everything look obvious at first glance, I'd be happy to add it to the real style guide (assuming we're happy with using roles for standardized formatting).

op-ct replied:
This is now implemented; you can see it in action at https://simp.readthedocs.io/en/simp-maint-example-role-formatting/contributors_guide/Documentation.html

op-ct commented:
> This is a LOT easier to deal with. How about we move the 'Sample' bits into the Developer Guide and get this through?
> 
> It would be great to have it all done at once but I don't know that this is realistic (either to do or to review) all at the same time so it may need to be incremental change starting with the User's Guide.

This is very true—the conversion process still takes time and decisions, even though the substitutions are simple.

(But when that times comes: I have some vim macros, if you're interested)

op-ct commented:
Added documentation:

* Typographical conventions: https://simp.readthedocs.io/en/simp-maint-example-role-formatting/getting_started_guide/Introduction.html
* Documentation & style guide: https://simp.readthedocs.io/en/simp-maint-example-role-formatting/contributors_guide/Documentation.html

The tables in both pages pull from the same data source to keep things DRY.

Also added superscript icons (ala wikipedia) to denote when hyperlinks link to an external site.

trevor-vaughan commented on docs/_extensions/simp_roles.py:
```diff
@@ -0,0 +1,74 @@
+      node = nodes.reference(rawtext, text, refuri=url, **options)
+      if re.search('/', text):
+        fixed_text = re.sub('/', '-', text)
+        print("WARNING: Use '-' to separate Puppet module names (.e.g., %s', not '%s')\n" % (fixed_text, text))
```
I think that this warning is misplaced and we should always use the `author/module_name` format.

Reasons:

* There doesn't appear to be any definitive guidance on this and I think that it's easier to read as `author/module_name`
* The forge lists modules as `author/module_name`
* The dependencies list in all modules that I've generally seen list as `author/module_name`

Differences:

* The Puppetfile appears to like `author-module_name`
* The `name` entry in the metadata.json needs to be `author-module_name`

So, personally, I'd like to come down on the side of 'easier to understand' with `author/module_name`.

lnemsick-simp replied:
Currently, the validation doesn't match the example.

op-ct replied:
@trevor-vaughan :+1:  A few years ago, Puppet started preferring `org-name` over `org/name`, as seen in metadata.json, Puppet commands, and (I think) the Forge API.  However, the most relevant guidance I've found has been the [official Puppet documentation on module _metadata_][1]:

> "**Tip**: Although the Forge expects to receive modules named `username-module`, its web interface presents them as `username/module`. Always use the `username-module` style in your metadata files and when issuing commands.

I'll still be glad to remove the warning—that metadata guidance doesn't explicitly apply to documentation nomenclature, and if there is such guidance Puppet is obviously not in a hurry to apply it to their own docs.  I personally prefer `org/name` because it makes the namespacing obvious and is less easy to mistake for an RPM or program name.


[1]: https://puppet.com/docs/puppet/6.17/modules_publishing.html#naming-your-module

op-ct replied:
Warning removed in latest commit; all valid delimiters are welcome in our docs!
